### ISSUE TITLE
Performance problem persists in 2019, with non-idiomatic but correct grammar; a question about algorithmic complexity

### ISSUE THREAD

**Issue by davewathaverford** [pegjs/pegjs#623]
### Issue type

- **Bug Report:** 
- **Feature Request:** 
- **Question:** 
- **Not an issue:** 

### Prerequisites

- **Can you reproduce the issue?:** 
- **Did you search the repository issues?:** 
- **Did you check the forums?:** 
- **Did you perform a web search (google, yahoo, etc)?:** 

### Steps to Reproduce

1. 
2. 
3. 

**Example code:**

```js
  // An example of code related to this issue
```

**Expected behavior:**

**Actual behavior:**

### Software

- **PEG.js:** 
- **Node.js:** 
- **NPM or Yarn:** 
- **Browser:** 
- **OS:** 
- **Editor:**

**Comment by meisl** [comment]
@davewathaverford super interesting observation!
I've got a very short answer for you. But I'd love to discuss it more, in particular what could be done *automatically* to alleviate the problem.

#### Use the results cache - boom! (it's NOT a bug per se)

The linear time guarantee only holds for packrat parsing, which means memoization. Without the "results cache" it's just simple recursive descent, and when backtracking occurs, everything down the line has to be computed again.
What's even worse: PEG has prioritized choice, which means you cannot just swap the alternatives in `addPrec` and `callPrec`.
However, the results cache comes at a cost: mainly memory consumption.

There are solutions to this, even without `+`, `*` or `?` (which are problematic too when it comes to packrat parsing).
But before that, let's try to understand the problem in bit more detail.

#### What I think is going on:
Let's look at the tree of recursive calls being made. `addPrec` and `callPrec` both have two alternatives (`primary` also has two, but that doesn't matter much in this case).
Assume the first alternative is represented by a left child, the second by a child on the right.
The `primary` nodes only have one child `addPrec`, except for the last but one level (at the "x"), where it first tries to match "(", and then `id`.
Bottom left part of it looks like so:
``` 

                                   /
                               prim
                                 |
                                add
                               /   \
                              /    ...
                             /
                         call
                        /    \
                       /     ...
                      /
                     /
                 prim
                  | 
                 add
                /    \
               /      \
              /        \
          call          call
          /  \         /   \
      prim   prim   prim   prim
       / \    / \    / \    / \
     "(" id "(" id "(" id "(" id
```  
The successful derivation of "(((...(x)...)))" is **the *rightmost* path** in this tree. However, since always the wrong alternative is explored first (which is unavoidable due to prioritized choice), **the whole tree is being walked**.
Let n be the number of pairs of parentheses. Then there are 3n+1 levels, but it bifurcates only at 2n+1 levels (think: `prim - add` as one node). That makes for total of 2**(2n+1) *paths* and (3n+1) * 2**(2n+1) *calls* - **exponential in n**.

I hope I got it correct so far. Was just thinking about it, haven't tried it (yet).

As said, I find this super interesting, and I'm going to post some pegjs here, that can be tried directly in https://pegjs.org/online . I'm thinking: why not just use the grammar itself for profiling this problem?

**Comment by davewathaverford** [comment]
Ah, this could really help simplify our project. I'll give it a try on my full example.

-- Dave W (sent from my phone, so please excuse brevity, speak-o's, and swype-o's)

**Comment by meisl** [comment]
Hehe, I bet whatever you want that memoization *does* help!
The theory is pretty old, but they couldn't afford the memory back then.

I've done a first sketch of my profiling idea, wich you'll find below.

I've simplified your grammar: names, omitted whitespace and id (just one id "x"),
so we can see what's really essential.
There's quite a bit of plumbing and preparation upfront, which will pay off later.
The actual grammar (variant) is short, starts after the `// -----------------`

Also: I think your grammar doesn't do what you really want:
* In `add` (originally `addPrec`), first alternative: I think you want `call "+" add` rather than `call + call`. For example `x+x+x` should be accepted, right?
* Then, so should `x(x)(x)`, maybe? (it's not accepted by my grammar either).

Now have fun trying it out. Watch in particular how the call count grows with increasing nesting depth. With and without (careful!) results cache.

```
// Deep Nesting: https://github.com/pegjs/pegjs/issues/623
{
  const tStart = Date.now();
  const calls = { add: 0, call: 0, prim: 0 };
}

start = profile:v1 // just to make it easy to switch between variants
/*
/  But: BE CAREFUL! You may easily choke your browser and loose your edits!
/  (with "results cache" you're on the safe side)
/
/  Variants:
/    - [ ] v0: using * operator
/    - [x] v1: variant from OP, choices with common prefix <<<<< BEWARE!
/    - [ ] v2: refactored v1, without * but ?
/    - [ ] v3: performant variant without *, + or ?
*/
{
  Object.assign(profile, {
    "max nesting": profile.n,
    "time (ms)": Date.now() - tStart,
    calls,
  });
  delete profile.n;

  return profile;
}

//v0 = p:add0 { p.variant = 0; return p; }
v1 = p:add1 { p.variant = 1; return p; }
//v2 = p:add2 { p.variant = 2; return p; }
//v3 = p:add3 { p.variant = 3; return p; }


bumpA = "" { calls.add++;  }
bumpC = "" { calls.call++; }
bumpP = "" { calls.prim++; }

// ------------------------------------------------------------------

add1 = bumpA p:(
       h:call1 "+" t:add1 { h.n = Math.max(h.n, t.n); return h; }
     / call1
  ) { return p; }

call1 = bumpC p:(
        h:prim1 "(" t:add1 ")" { t.n++; h.n = Math.max(h.n, t.n); return h; }
      / prim1
  ) { return p; }

prim1 = bumpP p:(
       "(" p:add1 ")" { p.n++; return p; }
	 / "x" { return {n: 0} }
  ) { return p; }
  

```

**Comment by davewathaverford** [comment]
Your demonstration makes sense to me; regarding the original grammar, yes, I do want x+x+x, but was submitting my bug report with the simplest grammar that created the problem.

**Comment by meisl** [comment]
Thanks :)
Re `x+x+x`: doesn't make much of a difference, was just wondering.
If you want me to explain how things work in my grammar - just ask (I am not perfect).
I'm trying to keep the plumbing down, but it may still be confusing.

I had more already, but lost quite some work because I switched to the pathological case with nesting ~17 but without memoization. That crashed the page and I lost it. Hence the "beware"... 

Gonna post a refactored version including variant 0 (`*`) soon. Results will be interesting.

**Comment by davewathaverford** [comment]
I've git-reverted back to a version of my grammar from before I started
doing unpleasant things to get better performance with pegjs, and I can
parse twenty-deep parenthesized x and all the examples from the grammar I
had started with, before beginning that digression into pegjs performance.
Feel free to push forward with your exploration if it interests you, but
I'm happy to move forward with my work based on what I'm seeing now ... I
just need to remember to turn on memoization each time!

Dave W

**Comment by polkovnikov-ph** [comment]
Just avoid choices that start with a common sequence.

There are implementation of PEGs that need explicit backtracking annotations (`try` in Parsec, for example) to enforce this rule. It's not like that in PEG.js probably to make its usage easier for less experienced people.

**Comment by davewathaverford** [comment]
Avoiding common sequences ("Left factoring") is fairly annoying, at least
in my experiences, when one or both choices start with (different) grammar
variables rather than literals. It also makes my grammar more and more
different from my abstract structure that I'm trying to parse. If there is
a way to make all of this easy, I'd appreciate a pointer. But, for now, I'm
happy with "turn on memoization", as that lets me write a grammar that
makes sense in terms of the parsed language, rather than having to think
about parsing details (yes, I'm more familiar with LR(1) and GLR parsers
and CFG's that PEGs, if you're wondering).

Dave W

**Comment by meisl** [comment]
@polkovnikov-ph, @davewathaverford: right, that's the problem. But: will the advice "no common prefixes" actually help less experienced ppl? Or, as @davewathaverford said, left factoring is *ugly* to do by hand.
PEG.js is already able to detect indirect left recursion, so why not also this?
And if so, rewrite / use different strategy.
*That* would make it easier, I think.

Right now I'm trying to get clear how hard such a thing would be really. Not done yet :)

--- 

So, @davewathaverford: Well, I will continue. At least up to all 4 variants I have in mind. v0 and v1 are only rebuilding your example, v2 and v3 will be actual refactorings of the grammar - which, I think, could be done *automatically*.
Always turning on the cache (or using `*` or `+`, for that matter) has its own disadvantages. Manual left-factoring - well, you've already mentioned that yourself.
Maybe you could just quickly try out my stuff for the time being, and possibly give me a note, when it doesn't work for you, or doesn't make sense?

Refactored profiling example incl. v0 follows.

---
p.s.: @davewathaverford don't worry, you're surely rather more than "less experienced". PEG is basically just plain recursive descent, but with *memoization*, and thus guaranteed linear time (if you don't screw up (-> `*`/`+`!). Also: unlimited lookahead, which you don't have with LL(k)/LR(k). But as always: "it depends". If you want, we can discuss your actual use-case, ie. how long, how deeply nested.

**Comment by meisl** [comment]
### Refactored version incl. variant 0 (using `*`)

#### v0 (using `*`) no cache *versus* v1 (OP) with cache
*Exactly* the same nr of calls ~> linear in the input length.
Note: that's only the calls we're actually measuring. There is a difference elsewhere, but we don't see it; even the more with such small a sample.
I suggest using this input (depth 10): `((((((((((x))))))))))`.
It's a bit unwieldy to switch between the two, you have to also change "Use results cache".
I'd like to detect if it's on and include that in the report, but I don't know how.

#### Another interesting effect
Take v1 *without* cache, and compare results for these two inputs:
* `((((((((((x))))))))))`
* `((((((((((x))))))))))+x`

***The latter is longer, but actually yields only about half the nr of calls. Can you tell why?***

---

```
// Deep Nesting: https://github.com/pegjs/pegjs/issues/623
{
  const tStart = Date.now();
  const calls = { add: 0, call: 0, prim: 0 };
  
  // Aggregate results from one or more profile objects.
  // head: profile object
  // tail: list of profile objects
  function listUpdate(head, tail = []) {
    head.n = tail.map(p => p.n).reduce(Math.max, head.n);
    return head;
  }
  // Nicify final output
  function polish(profile) {
    calls.TOTAL = calls.add + calls.call + calls.prim;
    Object.assign(profile, {
      "max nesting": profile.n,
      "time (ms)": Date.now() - tStart,
      calls,
    });
    delete profile.n;
    return profile;
  }
}
/*
/  Grammar variants (choose at start rule):
/    - [x] v0: using * operator
/    - [x] v1: choices with common prefix <<<<< BEWARE!
/    - [ ] v2: refactored v1, without * but using ?
/    - [ ] v3: performant variant without *, + or ?
/
/  BE CAREFUL! You may easily choke your browser and loose your edits!
/  (with "results cache" you're on the safe side)
*/

start = profile:v1 // <<<<<<<< SELECT VARIANT HERE <<<<<<<<
{ return polish(profile) }

v0 = p:add0 { p.variant = 0; return p; }
v1 = p:add1 { p.variant = 1; return p; }
//v2 = p:add2 { p.variant = 2; return p; }
//v3 = p:add3 { p.variant = 3; return p; }

bumpA = "" { calls.add++;  }
bumpC = "" { calls.call++; }
bumpP = "" { calls.prim++; }

// ------------------------------------------------------------------

add0 = bumpA p:(
        h:call0 t:("+" t:add0 { return t })*
          { t.n++; return listUpdate(h, t) }
      / call0
  ) { return p; }

call0 = bumpC p:(
        h:prim0 t:("(" t:add0 ")" { return t })*
          { t.n++; return listUpdate(h, t) }
      / prim0
  ) { return p; }

prim0 = bumpP p:(
       "(" p:add0 ")" { p.n++; return p; }
	 / "x" { return {n: 0} }
  ) { return p; }
  
// ------------------------------------------------------------------

add1 = bumpA p:(
       h:call1 "+" t:add1 { return listUpdate(h, [t]); }
     / call1
  ) { return p; }

call1 = bumpC p:(
        h:prim1 "(" t:add1 ")" { t.n++; return listUpdate(h, [t]); }
      / prim1
  ) { return p; }

prim1 = bumpP p:(
       "(" p:add1 ")" { p.n++; return p; }
	 / "x" { return {n: 0} }
  ) { return p; }
```

**Comment by meisl** [comment]
I've just forked pegjs and put above stuff into [tools/benchmark/nesting](https://github.com/meisl/pegjs/tree/master/tools/benchmark/nesting) (don't know better atm).
This gives me diffs, which I now need.

Will put subsequent examples there as well, and point to them from here.

**Comment by meisl** [comment]
@davewathaverford Re minimal example: I tried to really boil it down. I could just [link to my fork](https://github.com/meisl/pegjs/commit/62398ae7c8edf202bd28a3a81158aae7eb77d586?short_path=6b65b14#diff-6b65b14a591558d2d2f704032b9e7237), but I'll post the relevant part here again, for your convenience.

Do you think I got that right?

---

### Problem summary
[TODO: **common prefixes**, **memoization** (aka "results cache"), **automatic optimization?**]

Original problematic grammar boiled down:
```
A = C "+" A
  / C

C = P "(" A ")"
  / P

P = "(" A ")"
  / "x"
```
The problem as such could be reproduced with an even smaller grammar.
However, we'll stick with this, since we also want to discuss practical
aspects such as that the grammar should somewhat resemble the original intent.

Pathological case:
* *NO* memoization (ie. "Use results cache" OFF)
* AND deep nesting in inputs: eg depth 10 `((((((((((x))))))))))`
* BEWARE: parsing time will be **exponential in the nesting depth**
* Worth noting: there's a *huge* difference between *valid* and *invalid* input (both deeply nested). Eg.: `((((((((((x))))))))))` vs. `(((((((((())))))))))`. Guess what - the *latter takes much, much longer*! [TODO: why?!]

**Comment by davewathaverford** [comment]
Your boiled-down version of my code looks to me like it captures the
problem (I'm assuming you re-ran the test and confirmed the timing).

If you've been thinking about automatic optimization, it's not yet obvious
to me that there's a good motivation for it ... does the memoized version
really use a lot of space (i.e., worst case more than O(input), for a given
grammar; I've not checked)? If not, why not use *always* turn on
memoization (which was, perhaps, Ford's original plan) or make it the
default? CFG's can be parsed in O(input ^ 3) time, if my memory is correct,
so it's not clear to me why a PEG-based parser that's worst-case
exponential is particularly interesting.

And, if there's some reason to want non-memoized PEG parsers, and
optimization turns out to be hard, one option would be to just identify the
potential for exponential behavior and then provide a pop-up warning in
pegjs, recommending that memoization be used?

For now, even just changing the label "Use results cache" to "Use results
cache (leaving this off may cause exponential slow-down in some cases)", so
that folks don't dismiss pegjs.org as unworkably slow when they just didn't
realize that this button is the fix for that?

Dave W

**Comment by meisl** [comment]
Boiled-down version: of course I checked :)

Space cost: it *is* linear in the input size. But the constant factor can make it impractical: it is the nr of rules (assuming we make a function for each of them, which PEG.js does).
Basically, if you've got k rules and feed it n tokens of input, then you'd have to store n*k entries at most. Of course only entries that are really needed should be stored, and only *when* they're needed.
With this also comes an overhead (albeit small, and *constant*) for checking whether there is an entry already, and if not, to put it there when the parse function returns (aka "memoize the result").

> [...] so it's not clear to me why a PEG-based parser that's worst-case
> exponential is particularly interesting.

It isn't really.

>  [...] just identify the potential for exponential behavior and then provide a pop-up warning

In fact, that's where I want to go. But it *is* hard. It may very well be undecidable for PEGs, I'm not sure. However, try this in https://pegjs.org/online , you'll see:
```
     exp = prx / app           // "an expression is a primary expr. or an application"
     app = exp "(" exp ")"     // "an application is an expr. followed by expr. in parens"
     prx = "x" / "(" exp ")"   // "a primary expr. is an identifier or expr. in parens"
```
Btw.: this would accept `x(x)(x)`, as your `callPrec = primary  (_ "(" _ addPrec _ ")")*` would - but NOT your recursive version in the OP. This *is* a problem: we should not compare non-equivalent grammars.

Basically I just agree with you, that memoization should be the default.

I guess it's because of the [benchmarks](https://pegjs.org/development/benchmark), which are much more impressive without (~3x). Actually I want to understand why that is, and if something can be done about it.
As you have observed, it does pay off quickly for deeply nested structures *and* if the parsing of these is done with recursion alone. `*` or `+` sidestep this, they are implemented non-recursively.
But they also sidestep memoization, i.e. they're "doing secret things", that aren't stored in the results table. At least I suppose that they're implemented in such a fashion right now.

Maybe the generated parsers are "not lazy enough"?

**Comment by meisl** [comment]
Added variants 2 and 3 to [the example](https://github.com/meisl/pegjs/blob/master/tools/benchmark/nesting/nesting-profile.pegjs).

As mentioned above, I realized that v0 (with `*`) *is not equivalent* to the others, e.g. `x(x)(x)`. That's bad, even though the problem can be reproduced without this part of the language. 

After all, this is also about refactorings - which implies *no* change in semantics.
So next will be to look into how to re-establish the original semantics, while keeping the behaviour the same.

---

The profiling grammars are pretty much cluttered up with boilerplate, as to the point where it becomes difficult to spot the actual differences.
So I'll paste below all 4 grammars, but without the clutter.
* The `prim` rule is always the same, so we can concentrate on `add` and `call`
* Only v1 exposes the exponentional time, when NOT using the results cache
* v0 and v1 have (choices with) common prefixes, but not v2 and v3
* v0 accepts `x(x)(x)` while the others do not
* BUT: all four do accept `x+x+x`
* Hence: we were able to rebuild the `*` repetition for `add`, but not for `call`. Why?
* The sequence `"(" add ")"` occurs twice in all four. Maybe make an extra rule for that? ...this will eventually lead us to resurrecting the original semantics (~> "indirect left recursion", "left factoring")
* Finally: compare variants w.r.t. how they look with actual code inside them, ie. as real **P**arsing **E**xpression **G**rammars.

```
// v0: Repetition done with * (non-recursively), common prefixes:
//   add  = call ("+" add)*
//        / call
//   call = prim ("(" add ")")*
//        / prim
//   prim = "(" add ")"
//        / "x"

// v1: Repetition done via recursion, common prefixes:
//   add  = call ("+" add)
//        / call
//   call = prim ("(" add ")")
//        / prim
//   prim = "(" add ")"
//        / "x"

// v2: Avoid common prefixes by using ?:
//   add  = call ("+" add)?
//   call = prim ("(" add ")")?
//   prim = "(" add ")"
//        / "x"

// v3: Avoid common prefixes by using ε:
//   add  = call ("+" add / "")
//   call = prim ("(" add ")" / "")
//   prim = "(" add ")"
//        / "x"
```

**Comment by Mingun** [comment]
Just a note: PEG way to do this is:
```
add  = call ("+" call)* // note that use ("+" add) is useless here
call = prim group*
prim = group / "x"
group = "(" add ")"
```
Also, this grammar is strange in the rule `call`. The canonical way to parse expressions is:
```
expr = expr1 // whole expression

expr1 = expr2 op2 expr2
expr2 = expr3 op3 expr3
...
exprN = "(" expr1 ")"
```
where `opN` -- operator (not necessary infix, can be also postfix, prefix or even call form. For example, technically grouping in the last rule also is operator). Rules with highter numbers has highter precedence

**Comment by davewathaverford** [comment]
For that last "canonical way", note that the "add" grammar and "expr"
grammar are for different languages, I think. The first will parse x(x),
i.e., a function call, but I don't see a way to match x(x) in the "expr"
grammar. So, while the "expr" grammar allows nested parenthesized
expressions, it is not directly comparable to "add" grammar.

Dave W

**Comment by meisl** [comment]
Hi & Thanks, @Mingun
Your `group` rule is what I was hinting at above. And your grammar does accept mentioned `x(x)(x)` - with only *one* more rule, not *two* as I had in mind.

However:
The intention behind the `call` rule is function application, which, of course should have priority over any infix operator (like multiplication vs addition), and also be left-associative. Basically: we'd like to ultimately produce an AST like
```
  app
 /   \
fn   app
    /   \
   fn   arg
```
...for `x(x)(x)` ***and*** have that expressed in the grammar where one would expect it - namely as js code in the `call` rule. With `*` you get a - possibly empty - list. The differentiation between a primary value and an actual fn app then has to happen *in that code*, it's not anymore expressed in the grammar. The code also has to make it left-associative, again: not expressed in the grammar.
Just ask yourself: how would I actually make it from left to right, in the following "suggestive ASCII art (TM)" (again):
```
     call                    exp
    /    \             ~>     |
  h:x     t:[]                x


     call                    app
    /    \             ~>   /   \
  h:f     t:[x]            f     x


     call                       app
    /    \             ~>      /   \
  h:f     t:[x, y]           app    y
                            /   \
                           f     x


     call                          app
    /    \             ~>         /   \
  h:f     t:[x, y, z]           app    z
                               /   \
                             app    y
                            /   \
                           f     x
```

The quest is on to find some "sweet spot" how to best express such things as PEGs.

---

To both, @Mingun @davewathaverford: you're aware that function application *need not* be expressed by putting parentheses around the arguments?

Anyways, in case we do want to express it like that, then the parentheses can be viewed as some **strange hybrid of infix *and* circumfix operator**. Yes, I mean it:
* "circumfix" because it (the operator itself) consists of *two* symbols, that *surround* the stuff that is to be fed into the function
* "infix" because it has *on its left* the function that is to be applied
You notice: I was avoiding the term "argument" or "operand". That's because we're dealing with two levels here: a syntactic level - the `...(...)` operator and *its* operands, versus a semantic level - the function and *its* arguments*.

In this sense, the *parens-meaning-fn-application operator* is ***very different*** from the rather simple *parens-meaning-except-from-normal-precedence operator* (aka "group"). Syntactically, the former has *two* operands, while the latter only has *one*.

Hence: might be a good idea to have **different rules** for those two operators, even though they're expressed with the same set of terminals...

The "canonical" `expr` grammar from above is just for infix operators, nothing else. Not for prefix, postfix, circumfix, ternary (`... ? ... : ....`) - or our beloved *parens-meaning-fn-app* `...(...)`.
For those, infix only, however, it does define precedence cleanly. And all left-associative, btw.

...hmm, not really sure if any of that makes sense to you?

**Comment by meisl** [comment]
@Mingun I forgot:
Could you explain what you meant by
> `add  = call ("+" call)* // note that use ("+" add) is useless here`

?

**Comment by Mingun** [comment]
> With `*` you get a - possibly empty - list. 

This is not a problem. Just replace `*` with `+` in appropriate places.

> The differentiation between a primary value and an actual fn app then has to happen in that code, it's not anymore expressed in the grammar.

The code which you write in actions/predicates is also a part of PEG grammar. You cannot just drop it and tell that nothing changed. So you intentions _are expressed_, but it is not so explicit. The choice of this or that approach to parse of recursive structures -- just the choice in terms of performance. The PEG rules are very simple, and it is good! Just use repetition where repetitions are necessary. Do not use recursion for this! Recursion make a lot of backtracks, while a repetition not (just open https://pegjs.org/online with default example of arithmetical grammar and insert as many open brackets in the begin of input string, as you want. Even existence of _hundreds_ brackets does not influence performance a lot, as for valid, as for invalid (=mismatched count of brackets) input, _without cache_). And it is relative simple to build tree from these arrays, as I show in https://github.com/pegjs/pegjs/issues/592#issuecomment-442195640.

----
> @Mingun's "canonical" expr grammar - as given - is just for infix operators, nothing else.
Why not? Remember `C/C++` dereference operator:
```сpp
**x;// two dereferences
```
It is prefix operator.

Your not need to use repetitions in each level of precedence (for example, in function application). If you do not need repetitions, do not use it! But that not change general form, where you can have none, one or as many repetitions, as you want.

> And all left-associative, btw.

No. In my approach you have full control about associativity. The associativity makes sense only in the presence of a tree structure. As I showed in https://github.com/pegjs/pegjs/issues/592#issuecomment-442195640, you can build on a tree at the left or on the right, depending on needs.

> Maybe you could exemplify what you wrote in the comment, "not only for infix..."?

See my remark about dereference operator above.

----
> Could you explain what you meant by

First, notice, that you `v0` and `v2` grammars are equivalent -- repetitions in `v0` are never match more than one, because `add` in `("+" add)` part already consume all input.

Second, notice that your expressions in `v0` are excessive:
```
expr
  = expr something*
  / expr
```
Second leg in this grammar is never reachable, because `(...)*` always match, and if `expr` match, than first leg also match, and if `expr` not match, that second leg also not match. Of course, if you grammar have actions/predicates that maintain state, this is not true, but in that case you do not have such actions/predicates.

Third, make some substitutions in `v0` grammar:
1. ```
add = call ("+" add)*
    / call
```
2. As explained above, this is equivalent to
```
add = call ("+" add)*
```
3. Now substitute `add` in right hand:
```
add = call ("+" call ("+" add)*)*
```
4. As you see, if `add` match, it consume all input, so `*` can be replaced by `?` in grammar at step 2:
```
add = call ("+" call ("+" add)?)?
```
5. If you make backward substitution, you get your `v2` grammar:
add = call ("+" add)?
```
6. If you continue doing substitutions, you get
```
add = call ("+" call ("+" call (...()?...)?)?)?
```
Pattern `expr (expr (expr (...)?)?)?` can be replaced by `expr+` or `expr expr*`, _even if `expr` maintain state_.

----
Again, summary on a subject: from point of view of optimizations use the repeat primitives when you need repeat. Do not use recursion for this! Recursion make a lot of backtracks, while a repetition not. And it is relative simple to build tree from these arrays, as I show in https://github.com/pegjs/pegjs/issues/592#issuecomment-442195640

**Comment by meisl** [comment]
Hey @Mingun, big thanks for spotting `v0`/common prefix, that was my mistake: it should NOT have the choice in `add` and `call`. Got too busy with copy/paste, and somehow did not see it anymore. In my defence: it could be made clearer in the OP as well (@davewathaverford ?)
It also made me find another bug, now in the profiling code. Guess what - it was in processing arrays..!
Have corrected it above and in the example. Thanks again!

---
> First, notice, that you v0 and v2 grammars are equivalent

No, they are not. My https://github.com/pegjs/pegjs/issues/623#issuecomment-523127403 begins with:
> Added variants 2 and 3 to [the example](https://github.com/meisl/pegjs/blob/master/tools/benchmark/nesting/nesting-profile.pegjs).
> 
> As mentioned above, I realized that v0 (with `*`) *is not equivalent* to the others, e.g. `x(x)(x)`.

That is to say: paste [example](https://github.com/meisl/pegjs/blob/master/tools/benchmark/nesting/nesting-profile.pegjs) into https://pegjs.org/online and compare v0 with v2 on the input `x(x)(x)`.
Please do try it yourself. Look for `start = profile:v0 // <<<<<<<< SELECT VARIANT HERE <<<<<<<<` (~ line 36)

So: either there's something wrong with your explanation, or there's another flaw in my implementation - or even something wrong with PEG.js.
ATM I cannot see any of these...?!

Let's get this out of the way first.
Then, as said: all three v1, v2 and v3 must be made to accept `x(x)(x)`. 
Only after that we can make reasonable comparisons.

**Comment by meisl** [comment]
Re `x(x)(x)`: let me recap, starting with another citation from my https://github.com/pegjs/pegjs/issues/623#issuecomment-523127403 above (emphasis added):
> * v0 accepts x(x)(x) while the others do not
> * BUT: all four do accept x+x+x
> * **Hence: we were able to rebuild the * repetition for `add`, but not for `call`. Why?**
> * The sequence "(" add ")" occurs twice in all four. Maybe make an extra rule for that? ...this will eventually lead us to resurrecting the original semantics (~> "indirect left recursion", "left factoring")

Then we were talking about operators, in particular infix, circumfix and what I called "hybrid of infix and circumfix".

In `add` we have infix `"+"`, in `call` we have in&circumfix" `"... (" ... ")"`, aka "*parens-meaning-fn-application*".

So again: what exactly is it, that made the refactoring (attempt) in the OP work for `add` but fail for `call`?

**Comment by meisl** [comment]
@Mingun, you are aware that 
* the actual non-equivalence of v0 and v1 (at that time), was discovered [very early on in this thread](https://github.com/pegjs/pegjs/issues/623#issuecomment-522691966)
* right in the OP `*` is stated as NOT having a performance problem

?

**Comment by Mingun** [comment]
Yes, `v0` and `v2` are not equivalent, my mistake. Anyway, you `v0` grammar is excessive in the `add` rule: `("+" add)*` will must only 0 or 1 times, because:
1. It match when `"+"` match and `add` match
2. If `add` match, it consume all input, because
    1. it call `call` that consume some input
    2. it consume `"+"`
    3. `add` consume all input because see point 2

This kind of excessive can be easely detected when #488 will be implemented (but it needs to be read as "finding unattainable fragments of grammar," not just rules that are particular cases of fragments).

It just seemed to me that in the `call` the same pattern (`expr = something expr*`), but there it is different -- this is my mistake.

What `OP` mean?

**Comment by meisl** [comment]
First off, @Mingun: I really appreciate this discussion. Thanks! :smiley:

#### Re "OP":
By that I mean Opening Post - the [very first comment](https://github.com/pegjs/pegjs/issues/623#issue-480874026) with wich the issue was opened.
OP is also often used for referring to the person who wrote it; then you read it as Original Poster.

Btw: let me emphasize that `v0` is not *my* grammar. It is from the OP, and as such is part of the question - *as it was given there*.

**Comment by meisl** [comment]
I have not forgotten about the rest: precedence, what's "expressed" in a PEG.
But for now: only re your last post, @Mingun:

---

### Refactoring worked for `add`

You are right that `v0`'s `add` rule is "excessive". I'd just put it (=say that) differently:
`v0`'s `add` can be replaced by the `add` from `v1`, `v2` or `v3` - without changing the language.
Or: "*refactoring worked* for rule `add`".
Before we look at how to justify such a claim, let's recall
```
/* v0 */  add = call ("+" add)*

/* v1 */  add = call "+" add
              / call

/* v2 */  add = call ("+" add)?

/* v3 */  add = call ("+" add / "")
```
  * Your argument is a valid proof. It is an inductive proof (inductive =~ "recursive in point 2."). 
  * How about saying "*all four grammars accept `x`, `x+x`, `x+x+x`, `x+x+x+x`, `x+x+x+x+x`, ...*"?

The second is actual practice: *tests*. The "*...*" in there represents the fact that a 100% coverage will be impossible (here). So that is NOT a complete proof, whereas induction does handle the "*...*".

**Both perspectives must be considered, the theoretical (your argument) *and* the practical (tests).**
Remember Donald Knuth: "I have only proved it correct, didn't try it".
Just saying: both, you and the OP had overlooked that the refactoring worked for `add` but failed for `call`.
On the other hand: imagine we're testing only cases `x` and `x+x`. Well: replace the rhs ("right-hand-side") `add` in each by `call`. Then still all four will accept `x` and `x+x`, **but** `x+x+x`, ... would *only* be accepted by `v0`. Btw.: I did actually get trapped by this, but realized in time.

Thanks for mentioning #488. Sounds interesting, will look into it.

---

### Refactoring failed for `call`
Let's try to formulate the "recipee" that was applied to both, `add` and `call` in `v0` - but worked out fine only for `add`:
* if (in the grammar) there's a *sequence* of the form `prefix repd*`:
  - replace the sequence by a *choice* `prefix repd | prefix`

Well, in this form it should by obvious that such an algorithm is utterly flawed in many ways:
  * no say about the inner structure of `prefix` or `repd`
  * recipee not applied recursively to neither, `prefix` or `repd`
  * from that: ...

**Yet, this alone is enough to transform `v0` to `v1` - *and* it does work fine for the `add` rule!**

#### Then why not for `call`?
The only structural difference between `v0`'s `add` and `call`
```
add  = call ( "+" add     )*
call = prim ( "(" add ")" )*
```
...is **the suffix `")"` after the `add` in the `*`-repeated part**.

ATM ("at the moment") I am unable to give a general, exhaustive condition for "when to take extra care".
But for the grammar at hand, this is it.

---

### Why all those variants?
Well, for me it's about getting *really* clear about all that.
And, maybe, developing some ideas for automated optimization / guidance. In the spirit of #488.

Should it not already be obvious how actual code, and going step-by-step etc. helps me with that,
I'm totally ready to explain / open for discussion. Just ask.

**Comment by meisl** [comment]
I didn't get the [call tree thing](https://github.com/pegjs/pegjs/issues/623#issuecomment-522667361) quite right. That is, for inputs `x`, `(x)`, `((x))`, ...

For *n* being the number of pairs of parentheses in such an input, actually:
* the number of *paths* (= `id` leaves) is 4<sup>*n*+1</sup>
* the number of `add` nodes is (4<sup>*n*+1</sup> - 1)/3
* the total number of nodes, incl. `id` (=nr of calls) is 10(4<sup>*n*+1</sup> - 1)/3 + 1

or, without the `id` nodes (= `v1`, ie: rule `id` eliminated in [the example code](https://github.com/meisl/pegjs/blob/master/tools/benchmark/nesting/nesting-profile.pegjs)):
* the number of *paths* (= `prim` leaves) is 4<sup>*n*+1</sup>
* the number of `add` nodes is (4<sup>*n*+1</sup> - 1)/3
* the total number of nodes (=nr of calls) is 7(4<sup>*n*+1</sup> - 1)/3

@everyone: please feel invited to check me on this. I mean: counter-check it using the profiling code.

I'll update my fork and mentioned comment of mine, including derivations (of formulae). But it may take a while.
